evm/p2p/kademlia.py

```diff
@@ -0,0 +1,581 @@
+        # requests. It is ok for wait_ping() to timeout and return false here as that just means
+        # the remote remembers us.
+        yield from self.wait_ping(node)
+
+        logger.debug("bonding completed successfully with {}".format(node))
+        self.update_routing_table(node)
+        return True
+
+    @asyncio.coroutine
+    def bootstrap(self, bootstrap_nodes):
+        bonded = yield from asyncio.gather(*[self.bond(n) for n in bootstrap_nodes])
+        if bonded.count(True) == 0:
+            logger.info("Failed to bond with bootstrap nodes {}".format(bootstrap_nodes))
+            return
+        yield from self.lookup(self.this_node.id)
+
+    @asyncio.coroutine
+    def lookup(self, node_id):
+        """Lookup the closest nodes to the given node.
+        Lookup performs a network search for nodes close to the given target.
+
+        It approaches the target by querying nodes that are closer to it on each iteration.  The
+        given target does not need to be an actual node identifier.
+        """
+        nodes_asked = []
+        nodes_seen = []
+
+        @asyncio.coroutine
+        def _find_node(node_id, remote):
+            self.wire.send_find_node(remote, node_id)
+            candidates = yield from self.wait_neighbours(remote)
+            if len(candidates) == 0:
+                logger.info("got no candidates from {}, returning".format(remote))
+                return candidates
+            candidates = [c for c in candidates if c not in nodes_seen]
+            logger.info("got {} new candidates".format(len(candidates)))
+            # Add new candidates to nodes_seen so that we don't attempt to bond with failing ones
+            # in the future.
+            nodes_seen.extend(candidates)
+            bonded = yield from asyncio.gather(*[self.bond(c) for c in candidates])
+            logger.info("bonded with {} candidates".format(bonded.count(True)))
+            return [c for c in candidates if bonded[candidates.index(c)]]
+
+        def _exclude_if_asked(nodes):
+            nodes_to_ask = list(set(nodes).difference(nodes_asked))
+            return sort_by_distance(nodes_to_ask, node_id)[:k_find_concurrency]
+
+        closest = self.routing.neighbours(node_id)
+        logger.info("starting lookup; initial neighbours: {}".format(closest))
+        nodes_to_ask = _exclude_if_asked(closest)
+        while nodes_to_ask:
+            logger.info("node lookup; querying {}".format(nodes_to_ask))
+            nodes_asked.extend(nodes_to_ask)
+            results = yield from asyncio.gather(
+                *[_find_node(node_id, n) for n in nodes_to_ask])
+            [closest.extend(candidates) for candidates in results]
+            closest = sort_by_distance(closest, node_id)[:k_bucket_size]
+            nodes_to_ask = _exclude_if_asked(closest)
+
+        logger.info("lookup finished: {}".format(closest))
```

Comment:
This message should also probably contain the `node_id` that the lookup was performed on.